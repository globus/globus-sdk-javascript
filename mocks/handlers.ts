import { RestContext, RestRequest, rest } from "msw";

/**
 * A mirrored request from MSW. This can be used to easilty validate the shape
 * of a request that was dispatched to the server.
 */
export type MirroredRequest = {
  __msw: "FALLBACK";
  req: Pick<RestRequest, "url" | "method"> & {
    /**
     * MSW returns a `Headers` object, which is not JSON serializable. We
     * mirror as a plain object.
     */
    headers: {
      [key: string]: string;
    };
    /**
     * JSON payload provided with the request, if any.
     */
    json?: unknown;
  };
  ctx: RestContext;
};

export const handlers = [
  /**
   * For the majority of tests, we simply want to validate the shape of the
   * request object that was dispatched to the server. By returning the request
   * object, here, assertions can be run against it in individual tests.
   */
  rest.all("*", async (req, res, ctx) => {
    const headers: MirroredRequest["req"]["headers"] = {};
    req.headers.forEach((value, key) => {
      headers[key] = value;
    });
    const mirror: MirroredRequest = {
      __msw: "FALLBACK",
      req: {
        url: req.url,
        method: req.method,
        headers,
      },
      ctx,
    };

    try {
      mirror.req.json = await req.json();
    } catch (_e) {
      // ignore error
    }

    return res(ctx.json(mirror));
  }),
];
