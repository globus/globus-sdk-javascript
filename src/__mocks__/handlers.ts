import { HttpResponse, http } from 'msw';

/**
 * A mirrored request from MSW. This can be used to easilty validate the shape
 * of a request that was dispatched to the server.
 *
 * @deprecated Use the `mirror` method instead; `export` will be removed in a future release.
 */
export type MirroredRequest = {
  __msw: 'FALLBACK';
  req: Pick<Request, 'url' | 'method'> & {
    /**
     * MSW returns a `Headers` object, which is not JSON serializable. We
     * mirror as a plain object.
     */
    headers: Record<string, string>;
    /**
     * JSON payload provided with the request, if any.
     */
    json?: unknown;
    /**
     * Form data provided with the request, if any.
     */
    formData?: object;
  };
};

export const handlers = [
  /**
   * For the majority of tests, we simply want to validate the shape of the
   * request object that was dispatched to the server. By returning the request
   * object, here, assertions can be run against it in individual tests.
   */
  http.all('*', async ({ request }) => {
    const headers: MirroredRequest['req']['headers'] = {};

    Array.from(request.headers.entries()).forEach(([key, value]) => {
      headers[key] = value;
    });

    const formData: Record<any, any> = {};
    if (request.headers.get('content-type')?.includes('application/x-www-form-urlencoded')) {
      const rawFormData = await request.formData();
      rawFormData.forEach((value, key) => {
        formData[key] = value;
      });
    }

    const mirroredRequest: MirroredRequest = {
      __msw: 'FALLBACK',
      req: {
        url: request.url,
        method: request.method,
        headers,
        formData,
      },
    };

    try {
      mirroredRequest.req.json = await request.json();
    } catch {
      // ignore error
    }

    return HttpResponse.json(mirroredRequest);
  }),
];

/**
 * When `msw` is used in a test, the `mirror` method can be used
 * on the Response of a `fetch` to provide prope type information.
 *
 * @param response The `Response` object from a `fetch` call that `msw` intercepted.
 */
export async function mirror(response: Response): Promise<MirroredRequest> {
  return response.json();
}
