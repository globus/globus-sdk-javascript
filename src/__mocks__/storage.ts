/**
 * Creates an in-memory Storage mock.
 * @see https://developer.mozilla.org/en-US/docs/Web/API/Storage
 */
export class MockStorage implements Storage {
  store: {
    [key: string]: string;
  } = {};

  getItem = jest.fn().mockImplementation((k) => this.store[k]);

  setItem = jest.fn().mockImplementation((key, value) => {
    this.store[key] = `${value}`;
  });

  removeItem = jest.fn().mockImplementation((key) => {
    delete this.store[key];
  });

  clear = jest.fn().mockImplementation(() => {
    this.store = {};
  });

  get length() {
    return Object.keys(this.store).length;
  }

  key(n: number) {
    return Object.keys(this.store)[n];
  }
}

export function createStorageMock() {
  let mock = new MockStorage();
  /**
   * Ensure `Object.keys` calls behave similiar to real `Storage`.
   */
  mock = new Proxy(mock, {
    ownKeys: (target) => Object.keys(target.store),
    getOwnPropertyDescriptor: () => ({
      enumerable: true,
      configurable: true,
    }),
  });
  return mock;
}
