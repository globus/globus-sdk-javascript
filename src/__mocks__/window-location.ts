(function () {
  let URL_BAR = new URL('https://JEST_URL_BAR');
  Object.defineProperty(globalThis, 'window', {
    value: {
      location: {
        replace: jest.fn().mockImplementation((url) => {
          URL_BAR = new URL(url);
        }),
        set href(url) {
          URL_BAR = new URL(url);
        },
        get href() {
          return URL_BAR;
        },
      },
    },
    writable: true,
  });
})();
