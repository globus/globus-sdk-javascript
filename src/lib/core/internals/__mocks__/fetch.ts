const actual: typeof import('../fetch') = jest.requireActual('../fetch');

module.exports = {
  ...actual,
  fetch: jest.fn().mockResolvedValue({}),
};
