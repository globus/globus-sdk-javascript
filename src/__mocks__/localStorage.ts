import { createStorageMock } from './storage';

export function setInitialLocalStorageState(initialState: Record<string, string>) {
  Object.keys(initialState).forEach((key) => {
    globalThis.localStorage.setItem(key, initialState[key]);
  });
}

export function mockLocalStorage(initialState: Record<string, string> = {}) {
  const mock = createStorageMock();
  Object.defineProperty(globalThis, 'localStorage', {
    value: mock,
    writable: true,
  });
  if (initialState) {
    setInitialLocalStorageState(initialState);
  }
}
