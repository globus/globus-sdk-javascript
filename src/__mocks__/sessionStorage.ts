import { createStorageMock } from './storage';

export function mockSessionStorage() {
  Object.defineProperty(globalThis, 'sessionStorage', {
    value: createStorageMock(),
    writable: true,
  });
}
