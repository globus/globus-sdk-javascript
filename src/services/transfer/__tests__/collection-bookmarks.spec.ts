import { collectionBookmarks } from '../index';
import { mirror } from '../../../__mocks__/handlers';

describe('collectionBookmarks', () => {
  test('getAll', async () => {
    const {
      req: { url, method, headers },
    } = await mirror(await collectionBookmarks.getAll());
    expect({
      url,
      method,
      headers,
    }).toMatchSnapshot();
  });

  test('create', async () => {
    const {
      req: { url, method, headers, json },
    } = await mirror(
      await collectionBookmarks.create({
        payload: {
          name: 'My New Bookmark',
          endpoint_id: 'fake-endpoint-uuid',
          path: '/my/bookmark/path',
        },
      }),
    );
    expect({
      url,
      method,
      headers,
      json,
    }).toMatchSnapshot();
  });

  test('get', async () => {
    const {
      req: { url, method, headers },
    } = await mirror(await collectionBookmarks.get('my-bookmark-id'));
    expect({
      url,
      method,
      headers,
    }).toMatchSnapshot();
  });

  test('update', async () => {
    const {
      req: { url, method, headers, json },
    } = await mirror(
      await collectionBookmarks.update('my-bookmark-id', {
        payload: {
          name: 'New Name',
        },
      }),
    );
    expect({
      url,
      method,
      headers,
      json,
    }).toMatchSnapshot();
  });

  test('remove', async () => {
    const {
      req: { url, method, headers, json },
    } = await mirror(await collectionBookmarks.remove('my-bookmark-id'));
    expect({
      url,
      method,
      headers,
      json,
    }).toMatchSnapshot();
  });
});
