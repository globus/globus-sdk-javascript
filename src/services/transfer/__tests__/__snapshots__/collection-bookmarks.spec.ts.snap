// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectionBookmarks create 1`] = `
{
  "headers": {
    "accept": "*/*",
    "accept-encoding": "gzip,deflate",
    "connection": "close",
    "content-length": "88",
    "content-type": "application/json",
    "host": "transfer.api.globusonline.org",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
  },
  "json": {
    "endpoint_id": "fake-endpoint-uuid",
    "name": "My New Bookmark",
    "path": "/my/bookmark/path",
  },
  "method": "POST",
  "url": "https://transfer.api.globusonline.org/v0.10/bookmark",
}
`;

exports[`collectionBookmarks get 1`] = `
{
  "headers": {
    "accept": "*/*",
    "accept-encoding": "gzip,deflate",
    "connection": "close",
    "host": "transfer.api.globusonline.org",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
  },
  "method": "GET",
  "url": "https://transfer.api.globusonline.org/v0.10/bookmark/my-bookmark-id",
}
`;

exports[`collectionBookmarks getAll 1`] = `
{
  "headers": {
    "accept": "*/*",
    "accept-encoding": "gzip,deflate",
    "connection": "close",
    "host": "transfer.api.globusonline.org",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
  },
  "method": "GET",
  "url": "https://transfer.api.globusonline.org/v0.10/bookmark_list",
}
`;

exports[`collectionBookmarks remove 1`] = `
{
  "headers": {
    "accept": "*/*",
    "accept-encoding": "gzip,deflate",
    "connection": "close",
    "host": "transfer.api.globusonline.org",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
  },
  "json": undefined,
  "method": "DELETE",
  "url": "https://transfer.api.globusonline.org/v0.10/bookmark/my-bookmark-id",
}
`;

exports[`collectionBookmarks update 1`] = `
{
  "headers": {
    "accept": "*/*",
    "accept-encoding": "gzip,deflate",
    "connection": "close",
    "content-length": "19",
    "content-type": "application/json",
    "host": "transfer.api.globusonline.org",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
  },
  "json": {
    "name": "New Name",
  },
  "method": "PUT",
  "url": "https://transfer.api.globusonline.org/v0.10/bookmark/my-bookmark-id",
}
`;
